// @wazo/euc-plugins-sdk/lib/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVENT_MOBILE_HEADER = exports.EVENT_CHANGE_TOOLBAR_DISPLAY = exports.EVENT_ON_SWITCH_STACK_TENANT = exports.EVENT_ON_CONNECTED_TO_STACK = exports.EVENT_REMOVE_BANNER = exports.EVENT_DISPLAY_BANNER = exports.EVENT_UPDATE_BADGE = exports.EVENT_ON_NEW_SESSION = exports.EVENT_IGNORE_CALL = exports.EVENT_PARTICIPANT_LEAVE_ROOM = exports.EVENT_PARTICIPANT_JOIN_ROOM = exports.EVENT_USER_LEAVE_ROOM = exports.EVENT_USER_JOIN_ROOM = exports.EVENT_REMOVE_MODAL = exports.EVENT_DISPLAY_MODAL = exports.EVENT_RESET_NAVBAR_COLOR = exports.EVENT_CHANGE_NAVBAR_COLOR = exports.EVENT_DISPLAY_NOTIFICATION = exports.EVENT_RESET_SOUNDS = exports.EVENT_CONFIGURE_SOUNDS = exports.EVENT_STOP_CURRENT_SOUND = exports.EVENT_PLAY_HANGUP_SOUND = exports.EVENT_PLAY_DOUBLE_CALL_SOUND = exports.EVENT_PLAY_INCOMING_CALL_SOUND = exports.EVENT_PLAY_NEW_MESSAGE_SOUND = exports.EVENT_PLAY_PROGRESS_SOUND = exports.EVENT_WS_MESSAGE = exports.EVENT_ROUTE_CHANGE = exports.EVENT_ON_MEETING_CREATED = exports.EVENT_OPEN_SETTINGS = exports.EVENT_OPEN_MEETING_LOBBY = exports.EVENT_CREATE_MEETING = exports.EVENT_OPEN_LINK = exports.EVENT_ON_CALL_HUNG_UP = exports.EVENT_ON_CALL_ACCEPTED = exports.EVENT_ON_CALL_ANSWERED = exports.EVENT_ON_CALL_MADE = exports.EVENT_ON_CALL_INCOMING = exports.EVENT_START_CALL = exports.EVENT_OPEN_LEFT_PANEL = exports.EVENT_CLOSE_LEFT_PANEL = exports.EVENT_ON_BACKGROUND_MESSAGE = exports.EVENT_ON_IFRAME_MESSAGE = exports.EVENT_SEND_BACKGROUND_MESSAGE = exports.EVENT_SEND_IFRAME_MESSAGE = exports.EVENT_ON_LOGOUT = exports.EVENT_APP_INITIALIZE = exports.EVENT_PLUGIN_UNLOADED = exports.EVENT_APP_UNLOADED = exports.EVENT_ON_LOADED = void 0;
exports.EVENT_MOBILE_ON_HEADER_BACK = exports.EVENT_MOBILE_ON_MENU_ACTION = exports.EVENT_MOBILE_SHOW_BOTTOM_NAV = exports.EVENT_MOBILE_CONTEXTUAL_MENU = void 0;
// Global
exports.EVENT_ON_LOADED = 'wazo/ON_LOADED';
exports.EVENT_APP_UNLOADED = 'wazo/EVENT_APP_UNLOADED';
exports.EVENT_PLUGIN_UNLOADED = 'wazo/EVENT_PLUGIN_UNLOADED';
exports.EVENT_APP_INITIALIZE = 'wazo/EVENT_APP_INITIALIZE';
exports.EVENT_ON_LOGOUT = 'wazo/EVENT_ON_LOGOUT';
exports.EVENT_SEND_IFRAME_MESSAGE = 'wazo/EVENT_SEND_IFRAME_MESSAGE';
exports.EVENT_SEND_BACKGROUND_MESSAGE = 'wazo/EVENT_SEND_BACKGROUND_MESSAGE';
exports.EVENT_ON_IFRAME_MESSAGE = 'wazo/EVENT_ON_IFRAME_MESSAGE';
exports.EVENT_ON_BACKGROUND_MESSAGE = 'wazo/EVENT_ON_BACKGROUND_MESSAGE';
// WDA
exports.EVENT_CLOSE_LEFT_PANEL = 'wazo/EVENT_CLOSE_LEFT_PANEL';
exports.EVENT_OPEN_LEFT_PANEL = 'wazo/EVENT_OPEN_LEFT_PANEL';
exports.EVENT_START_CALL = 'wazo/START_CALL';
exports.EVENT_ON_CALL_INCOMING = 'wazo/EVENT_ON_CALL_INCOMING';
exports.EVENT_ON_CALL_MADE = 'wazo/EVENT_ON_CALL_MADE';
exports.EVENT_ON_CALL_ANSWERED = 'wazo/EVENT_ON_CALL_ANSWERED';
exports.EVENT_ON_CALL_ACCEPTED = 'wazo/EVENT_ON_CALL_ACCEPTED';
exports.EVENT_ON_CALL_HUNG_UP = 'wazo/EVENT_ON_CALL_HUNG_UP';
exports.EVENT_OPEN_LINK = 'wazo/EVENT_OPEN_LINK';
exports.EVENT_CREATE_MEETING = 'wazo/EVENT_CREATE_MEETING';
exports.EVENT_OPEN_MEETING_LOBBY = 'wazo/EVENT_OPEN_MEETING_LOBBY';
exports.EVENT_OPEN_SETTINGS = 'wazo/EVENT_OPEN_SETTINGS';
exports.EVENT_ON_MEETING_CREATED = 'wazo/EVENT_ON_MEETING_CREATED';
exports.EVENT_ROUTE_CHANGE = 'wazo/EVENT_ROUTE_CHANGE';
exports.EVENT_WS_MESSAGE = 'wazo/EVENT_WS_MESSAGE';
exports.EVENT_PLAY_PROGRESS_SOUND = 'wazo/EVENT_PLAY_PROGRESS_SOUND';
exports.EVENT_PLAY_NEW_MESSAGE_SOUND = 'wazo/EVENT_PLAY_NEW_MESSAGE_SOUND';
exports.EVENT_PLAY_INCOMING_CALL_SOUND = 'wazo/EVENT_PLAY_INCOMING_CALL_SOUND';
exports.EVENT_PLAY_DOUBLE_CALL_SOUND = 'wazo/EVENT_PLAY_DOUBLE_CALL_SOUND';
exports.EVENT_PLAY_HANGUP_SOUND = 'wazo/EVENT_PLAY_HANGUP_SOUND';
exports.EVENT_STOP_CURRENT_SOUND = 'wazo/EVENT_STOP_CURRENT_SOUND';
exports.EVENT_CONFIGURE_SOUNDS = 'wazo/EVENT_CONFIGURE_SOUNDS';
exports.EVENT_RESET_SOUNDS = 'wazo/EVENT_RESET_SOUNDS';
exports.EVENT_DISPLAY_NOTIFICATION = 'wazo/EVENT_DISPLAY_NOTIFICATION';
exports.EVENT_CHANGE_NAVBAR_COLOR = 'wazo/EVENT_CHANGE_NAVBAR_COLOR';
exports.EVENT_RESET_NAVBAR_COLOR = 'wazo/EVENT_RESET_NAVBAR_COLOR';
exports.EVENT_DISPLAY_MODAL = 'wazo/EVENT_DISPLAY_MODAL';
exports.EVENT_REMOVE_MODAL = 'wazo/EVENT_REMOVE_MODAL';
exports.EVENT_USER_JOIN_ROOM = 'wazo/EVENT_USER_JOIN_ROOM';
exports.EVENT_USER_LEAVE_ROOM = 'wazo/EVENT_USER_LEAVE_ROOM';
exports.EVENT_PARTICIPANT_JOIN_ROOM = 'wazo/EVENT_PARTICIPANT_JOIN_ROOM';
exports.EVENT_PARTICIPANT_LEAVE_ROOM = 'wazo/EVENT_PARTICIPANT_LEAVE_ROOM';
exports.EVENT_IGNORE_CALL = 'wazo/EVENT_IGNORE_CALL';
exports.EVENT_ON_NEW_SESSION = 'wazo/EVENT_ON_NEW_SESSION';
exports.EVENT_UPDATE_BADGE = 'wazo/EVENT_UPDATE_BADGE';
exports.EVENT_DISPLAY_BANNER = 'wazo/EVENT_DISPLAY_BANNER';
exports.EVENT_REMOVE_BANNER = 'wazo/EVENT_REMOVE_BANNER';
// Portal
exports.EVENT_ON_CONNECTED_TO_STACK = 'wazo/EVENT_ON_CONNECTED_TO_STACK';
exports.EVENT_ON_SWITCH_STACK_TENANT = 'wazo/EVENT_ON_SWITCH_STACK_TENANT';
exports.EVENT_CHANGE_TOOLBAR_DISPLAY = 'wazo/EVENT_CHANGE_TOOLBAR_DISPLAY';
// Mobile
exports.EVENT_MOBILE_HEADER = 'wazo/mobile/EVENT_MOBILE_HEADER';
exports.EVENT_MOBILE_CONTEXTUAL_MENU = 'wazo/mobile/EVENT_MOBILE_CONTEXTUAL_MENU';
exports.EVENT_MOBILE_SHOW_BOTTOM_NAV = 'wazo/mobile/EVENT_MOBILE_SHOW_BOTTOM_NAV';
exports.EVENT_MOBILE_ON_MENU_ACTION = 'wazo/mobile/EVENT_MOBILE_ON_MENU_ACTION';
exports.EVENT_MOBILE_ON_HEADER_BACK = 'wazo/mobile/EVENT_MOBILE_ON_HEADER_BACK';
